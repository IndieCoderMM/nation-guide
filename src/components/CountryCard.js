import React from 'react';
import { Link } from 'react-router-dom';

const CountryCard = ({ name, area, flags, className }) => {
  return (
    <Link to={`/country/${name.common}`} key={area} className={className}>
      <div className="frame">
        <img src={flags.png} alt={flags.alt} />
      </div>
      <div>
        <h3>{name.common}</h3>
        <p>{area}</p>
      </div>
    </Link>
  );
};

export default CountryCard;
