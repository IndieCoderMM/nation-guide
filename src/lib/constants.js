import { FiGithub, FiLinkedin } from 'react-icons/fi';
import { BiGlobe } from 'react-icons/bi';

export const sortingOptions = [
  { label: 'Area (asc)', value: 'area-asc' },
  { label: 'Area (desc)', value: 'area-desc' },
  { label: 'Name (asc)', value: 'name-asc' },
  { label: 'Name (desc)', value: 'name-desc' },
];

export const socialLinks = [
  {
    href: 'https://github.com/indiecodermm',
    label: 'GitHub',
    icon: <FiGithub />,
  },
  {
    href: 'https://linkedin.com/in/hthantoo',
    label: 'LinkedIn',
    icon: <FiLinkedin />,
  },
  {
    href: 'https://heinthantoo.me',
    label: 'Portfolio',
    icon: <BiGlobe />,
  },
];
