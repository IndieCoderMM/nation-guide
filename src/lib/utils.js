import { SORTING_OPTIONS } from './constants';

/**
 * Generate a slug from a string.
 *
 * @param {string} name
 * @returns {string}
 */
export const generateSlug = (name) => name.toLowerCase().replace(/\s+/g, '-');

/**
 * Filter and sort a list of countries based on search query and sorting criteria.
 *
 * @param {string} query - The search query to filter countries by name.
 * @param {string} sorter - The sorting criteria ('area-d', 'area-a', 'name-d', 'name-a').
 * @param {Array} data - An array of country objects.
 * @returns {Array} - The filtered and sorted list of countries.
 */
export const filterAndSortCountries = (query, sorter, data) => {
  const validSorters = Object.values(SORTING_OPTIONS);
  if (!validSorters.includes(sorter)) {
    throw new Error('Invalid sorter');
  }

  const getName = (country) => country.name.common.toLowerCase();

  const filteredCountries = query.trim()
    ? data.filter((country) => getName(country).includes(query.toLowerCase()))
    : data;

  const sortedCountries = [...filteredCountries];

  if (sorter === SORTING_OPTIONS.AREA_DESC) {
    sortedCountries.sort((a, b) => b.area - a.area);
  } else if (sorter === SORTING_OPTIONS.AREA_ASC) {
    sortedCountries.sort((a, b) => a.area - b.area);
  } else if (sorter === SORTING_OPTIONS.NAME_DESC) {
    sortedCountries.sort((a, b) => a.name.common.localeCompare(b.name.common));
  } else if (sorter === SORTING_OPTIONS.NAME_ASC) {
    sortedCountries.sort((a, b) => b.name.common.localeCompare(a.name.common));
  }

  return sortedCountries;
};
