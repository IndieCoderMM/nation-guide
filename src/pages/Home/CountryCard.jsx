import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import { generateSlug } from '../../lib/utils';

import styles from './styles/CountryCard.module.css';

// eslint-disable-next-line object-curly-newline
const CountryCard = ({ name, area, flag, capital, population }) => {
  const slug = generateSlug(name);
  return (
    <div className={styles.card}>
      <Link
        to={`/country/${slug}`}
        key={area}
        aria-label={`Go to ${name} page`}
        className={styles.link}
      />
      <div className={styles.frame}>
        <img src={flag.png} alt={flag.alt} />
      </div>
      <section className={styles.body}>
        <header>
          <h2 className={styles.name}>{name}</h2>
        </header>
        <div className={styles.infoBox}>
          <div>
            <h3>Capital: </h3>
            <p>{capital}</p>
          </div>
          <div>
            <h3>Area: </h3>
            <p>{area.toString().concat(' kmÂ²')}</p>
          </div>
          <div>
            <h3>Population: </h3>
            <p>{population}</p>
          </div>
        </div>
      </section>
    </div>
  );
};

CountryCard.propTypes = {
  name: PropTypes.string.isRequired,
  area: PropTypes.number.isRequired,
  flag: PropTypes.shape({
    png: PropTypes.string.isRequired,
    alt: PropTypes.string.isRequired,
  }).isRequired,
  capital: PropTypes.string.isRequired,
  population: PropTypes.number.isRequired,
};

export default CountryCard;
