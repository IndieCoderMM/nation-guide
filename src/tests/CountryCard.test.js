import React from 'react';
import { MemoryRouter as Router, useLocation } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import CountryCard from '../components/CountryCard';
import sampleResponseData from './sampleResponseData';
import { generateSlug } from '../lib/utils';

const testCountry = sampleResponseData[0];

describe('CountryCard Component', () => {
  it('renders the component with correct data', () => {
    // eslint-disable-next-line object-curly-newline
    const { name, area, capital, population, flag } = testCountry;
    const { asFragment } = render(
      <Router>
        <CountryCard
          name={name.common}
          area={area}
          flag={flag}
          capital={capital}
          population={population}
        />
      </Router>,
    );

    // Check with snapshot
    expect(asFragment()).toMatchSnapshot();

    // Check if the data is displayed
    expect(screen.getByText(name.common)).toBeInTheDocument();
    expect(screen.getByText(`${area} kmÂ²`)).toBeInTheDocument();
    expect(screen.getByText(capital)).toBeInTheDocument();
    expect(screen.getByText(population)).toBeInTheDocument();
    expect(screen.getByAltText(flag.alt)).toHaveAttribute('src', flag.png);
  });

  it('renders a link to the country page with the correct slug', () => {
    // eslint-disable-next-line object-curly-newline
    const { name, area, capital, population, flag } = testCountry;
    render(
      <Router>
        <CountryCard
          name={name.common}
          area={area}
          flag={flag}
          capital={capital}
          population={population}
        />
      </Router>,
    );

    // Check if the link is rendered with the correct slug
    const link = screen.getByRole('link', {
      name: `Go to ${name.common} page`,
    });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute(
      'href',
      `/country/${generateSlug(name.common)}`,
    );
  });

  it('navigate to the country page when the link is clicked', () => {
    // Create a component that displays the current location
    const DisplayLocation = () => {
      const location = useLocation();
      return <div data-testid="route-location">{location.pathname}</div>;
    };
    // eslint-disable-next-line object-curly-newline
    const { name, area, capital, population, flag } = testCountry;
    render(
      <Router initialEntries={['/']} initialIndex={0}>
        <CountryCard
          name={name.common}
          area={area}
          flag={flag}
          capital={capital}
          population={population}
        />
        <DisplayLocation />
      </Router>,
    );

    // Click the link
    const link = screen.getByRole('link', {
      name: `Go to ${name.common} page`,
    });
    fireEvent.click(link);

    // Verify that location has changed to the expected country page URL
    const routeLocation = screen.getByTestId('route-location');
    expect(routeLocation.textContent).toBe(
      `/country/${generateSlug(name.common)}`,
    );
  });
});
