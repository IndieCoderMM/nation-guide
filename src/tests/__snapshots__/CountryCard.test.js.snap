// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryCard Component renders the component with correct data 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <a
      aria-label="Go to Japan page"
      class="link"
      href="/country/japan"
    />
    <div
      class="frame"
    >
      <img
        alt="The flag of Japan features a crimson-red circle at the center of a white field."
        src="https://flagcdn.com/w320/jp.png"
      />
    </div>
    <section
      class="body"
    >
      <header>
        <h2
          class="name"
        >
          Japan
        </h2>
      </header>
      <div
        class="infoBox"
      >
        <div>
          <h3>
            Capital: 
          </h3>
          <p>
            Tokyo
          </p>
        </div>
        <div>
          <h3>
            Area: 
          </h3>
          <p>
            377930 kmÂ²
          </p>
        </div>
        <div>
          <h3>
            Population: 
          </h3>
          <p>
            125836021
          </p>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
